File: sdk/typescript/examples/paginate.ts
import { createClient, paginate } from "../src";

const client = createClient();

// Offset/limit pattern — most rocksky endpoints work this way.
type Scrobble = { uri: string; track: { title: string; artist: string } };
type ScrobblesResponse = { scrobbles: Scrobble[] };

const iter = paginate<Scrobble>({
  fetch: async ({ limit, offset }) => {
    const page = await client.actor.getActorScrobbles<ScrobblesResponse>({
      did: "tsiry.bsky.social",
      limit,
      offset,
    });
    return page.scrobbles ?? [];
  },
  pageSize: 50,
  maxItems: 200,
});

for await (const s of iter) {
  console.log(s.track.title, "—", s.track.artist);
}

// Cursor pattern — getFollowers/getFollows.
type Follower = { did: string; handle: string };
type FollowersResponse = { followers: Follower[]; cursor?: string };

const followers = await client
  .paginate<Follower>({
    fetch: async ({ limit, cursor }) => {
      const page = await client.graph.getFollowers<FollowersResponse>({
        actor: "tsiry.bsky.social",
        limit,
        cursor,
      });
      return { items: page.followers, cursor: page.cursor };
    },
    pageSize: 100,
  })
  .toArray();

console.log("total followers:", followers.length);

File: sdk/elixir/examples/paginate.exs
# Walk through every page of an actor's scrobbles using Stream.unfold.
#
#   ROCKSKY_DID=alice.bsky.social mix run examples/paginate.exs

did = System.get_env("ROCKSKY_DID") || raise "set ROCKSKY_DID"
client = Rocksky.new(base_url: System.get_env("ROCKSKY_BASE_URL", "https://api.rocksky.app"))

page_size = 50

scrobbles_stream =
  Stream.unfold(0, fn offset ->
    case Rocksky.Actor.get_actor_scrobbles(client,
           did: did,
           limit: page_size,
           offset: offset
         ) do
      {:ok, %{"scrobbles" => []}} ->
        nil

      {:ok, %{"scrobbles" => batch}} ->
        {batch, offset + length(batch)}

      {:error, err} ->
        IO.warn("page #{offset}: " <> Exception.message(err))
        nil
    end
  end)
  |> Stream.flat_map(& &1)

count =
  scrobbles_stream
  |> Stream.take(500)
  |> Enum.reduce(0, fn _s, acc -> acc + 1 end)

IO.puts("Walked #{count} scrobbles for #{did}.")
